src/jdk.compiler/share/classes/com/sun/tools/javac/code/Types.java

```diff
@@ -5009,6 +5009,54 @@ public Type visitCapturedType(CapturedType t, S s) {
     }
     // </editor-fold>
 
+    // <editor-fold defaultstate="collapsed" desc="Unconditionality">
+    /** Check unconditionality between any combination of reference or primitive types.
+     *
+     *  Rules:
+     *  - widening from one reference type to another,
+     *  - boxing.
+     *
+     *  @param source     Source primitive or reference type
+     *  @param target     Target primitive or reference type
+     */
+    public boolean checkUnconditionallyExact(Type source, Type target) {
+        if (isSameType(source, target)) {
+            return true;
+        }
+
+        if (target.isPrimitive()) {
+            return (source.isReference() && isSubtype(boxedTypeOrType(erasure(source)), target)) ||
+                    (source.isReference() && isExactPrimitiveWidening(unboxedType(source), target) ) ||
+                    isExactPrimitiveWidening(source, target);
+        } else {
+            return isSubtype(boxedTypeOrType(erasure(source)), target);
+        }
+    }
+
+    /** Check unconditionality between primitive types.
+     *
+     *  - widening from one integral type to another,
+     *  - widening from one floating point type to another,
+     *  - widening from byte, short, or char to a floating point type,
+     *  - widening from int to double.
+     *
+     *  @param source     Source primitive type
+     *  @param target     Target primitive type
+     */
+    public boolean isExactPrimitiveWidening(Type source, Type target) {
+        if (isSameType(source, target)) {
+            return true;
+        }
+
+        return (source.isPrimitive() && target.isPrimitive()) &&
+                ((source.hasTag(BYTE) && !target.hasTag(CHAR) ||
```

Comment:
Does this mean that `byte` -> `char` is not exact? Aren't both integral types, thus invalidating `widening from one integral type to another` ?

Reply:
Indeed it is not unconditionally exact. byte to char combines both widening and narrowing primitive conversions 

https://docs.oracle.com/javase/specs/jls/se8/html/jls-5.html#jls-5.1.4

Reply:
Ha! Of course - the problem is not the definition of "integral type" (which includes `char`), but the fact that there's no widening from `byte` to `char`. I agree that the code (and the spec changes) do the right thing.